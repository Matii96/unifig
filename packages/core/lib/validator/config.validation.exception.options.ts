import { ValidationError } from 'class-validator';
import { Type } from '../utils/type.interface';

export interface ConfigValidationExceptionOptions {
  template: Type<any>;
  errors: ValidationError[];
}
