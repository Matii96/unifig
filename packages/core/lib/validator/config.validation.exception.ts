export class ConfigValidationException<TConfig extends object> extends Error {
  constructor(config: TConfig, msg: string) {
    super(`Validation failed for ${config.constructor.name}:${msg}`);
    this.name = ConfigValidationException.name;
  }
}
