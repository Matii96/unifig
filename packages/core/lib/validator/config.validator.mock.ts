export class ConfigValidatorMock {
  validate = jest.fn();
}
