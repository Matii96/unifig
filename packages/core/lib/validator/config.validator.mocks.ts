import { ValidationError } from 'class-validator';
import { DbConfigMock, TemplateMock } from '../core.mocks';

export const mockFailedValidation: ValidationError[] = [
  {
    target: new TemplateMock(),
    property: 'port',
    children: [] as any[],
    constraints: { isInt: 'port must be an integer number' },
  },
  {
    target: new TemplateMock(),
    value: {},
    property: 'db',
    children: [
      {
        target: new DbConfigMock(),
        property: 'url',
        children: [] as any[],
        constraints: { isString: 'url must be a string' },
      },
      {
        target: new DbConfigMock(),
        property: 'password',
        children: [],
        constraints: { isString: 'password must be a string' },
      },
    ],
  },
  {
    target: new TemplateMock(),
    value: {},
    property: 'db2',
    children: [
      {
        target: new DbConfigMock(),
        property: 'url',
        children: [] as any[],
        constraints: { isDefined: 'url must be a defined', isString: 'url must be a string' },
      },
      {
        target: new DbConfigMock(),
        property: 'subdb',
        children: [
          {
            target: new DbConfigMock(),
            property: 'url',
            children: [] as any[],
            constraints: { isString: 'url must be a string' },
          },
          {
            target: new DbConfigMock(),
            property: 'password',
            children: [],
            constraints: { isString: 'password must be a string' },
          },
        ],
        constraints: { isString: 'password must be a string' },
      },
    ],
  },
];
