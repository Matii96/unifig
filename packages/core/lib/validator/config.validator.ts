import { validateSync, ValidationError } from 'class-validator';
import { ConfigValidationException } from './config.validation.exception';

export class ConfigValidator {
  validate<TConfig extends object>(config: TConfig) {
    const errors = validateSync(config, { skipMissingProperties: false });
    if (errors.length === 0) {
      return;
    }
    const errorFormatted = '\n' + errors.map((error) => this.format(error)).join('');
    throw new ConfigValidationException(config, errorFormatted);
  }

  private format(error: ValidationError, parentPath = '') {
    let msg = error.constraints
      ? ` - ${parentPath}${error.property} has failed the following constraints:` +
        ` ${Object.keys(error.constraints).join(', ')}.` +
        ` Current value: ${error.value}\n`
      : '';
    msg += error.children?.map((child) => this.format(child, parentPath + `${error.property}.`)).join('');
    return msg;
  }
}
