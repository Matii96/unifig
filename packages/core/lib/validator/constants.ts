export const SECRET_PROPERTY_METADATA = Symbol('unifigSecretProperty');
