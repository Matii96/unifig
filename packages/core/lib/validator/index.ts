export * from './decorators';
export * from './errors';
