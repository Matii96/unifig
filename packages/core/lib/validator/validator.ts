import { ConfigValidationError } from './errors/validation.error';

export interface Validator {
  validate(configs: object[]): ConfigValidationError | undefined;
}
