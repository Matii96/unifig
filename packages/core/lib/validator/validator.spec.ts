import 'reflect-metadata';
import { DbConfigMock, TemplateMock } from '../core.mocks';
import { ConfigValidationException } from './errors/validation.exception';
import { ConfigValidator } from './validator.impl';
import { Validator } from './validator';

describe('ConfigValidator', () => {
  let validator: Validator;

  beforeEach(() => {
    validator = new ConfigValidator();
  });

  it('should return ConfigValidationException', () => {
    const config = new TemplateMock();
    expect(validator.validate([config])).toBeInstanceOf(ConfigValidationException);
  });

  it('should structure validation errors', () => {
    const config = new TemplateMock();
    config.db = new DbConfigMock();
    const validationResult = validator.validate([config])!;
    expect(validationResult.errors).toMatchSnapshot();
  });
});
