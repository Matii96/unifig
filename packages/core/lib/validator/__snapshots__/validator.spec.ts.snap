// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigValidator should structure validation errors 1`] = `
[
  {
    "errors": [
      {
        "currentValue": undefined,
        "failedConstraints": [
          {
            "details": "port must be an integer number",
            "name": "isInt",
          },
        ],
        "property": "port",
        "source": "PORT",
      },
      {
        "children": [
          {
            "currentValue": undefined,
            "failedConstraints": [
              {
                "details": "url must be a string",
                "name": "isString",
              },
            ],
            "property": "url",
            "source": "DB_URL",
          },
          {
            "currentValue": undefined,
            "failedConstraints": [
              {
                "details": "password must be a string",
                "name": "isString",
              },
            ],
            "property": "password",
            "source": "DB_PASSWORD",
          },
        ],
        "failedConstraints": undefined,
        "property": "db",
      },
    ],
    "template": [Function],
  },
]
`;
